src/Mvc/test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/HtmlHelperComponentsExtensionsTests.cs

```diff
@@ -0,0 +1,330 @@
+                    builder.AddContent(s++, "Hello world!");
+                    builder.CloseElement();
+                });
+                return Task.CompletedTask;
+            }
+        }
+
+        private class GreetingComponent : ComponentBase
+        {
+            [Parameter] public string Name { get; set; }
+
+            protected override void OnParametersSet()
+            {
+                base.OnParametersSet();
+            }
+
+            protected override void BuildRenderTree(RenderTreeBuilder builder)
+            {
+                var s = 0;
+                base.BuildRenderTree(builder);
+                builder.OpenElement(s++, "p");
+                builder.AddContent(s++, $"Hello {Name}!");
+                builder.CloseElement();
+            }
+        }
+
+        private class AsyncComponent : ComponentBase
+        {
+            private static WeatherRow[] _weatherData = new[]
+            {
+                new WeatherRow
+                {
+                    DateFormatted = "06/05/2018",
+                    TemperatureC = 1,
+                    Summary = "Freezing",
+                    TemperatureF = 33
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "07/05/2018",
+                    TemperatureC = 14,
+                    Summary = "Bracing",
+                    TemperatureF = 57
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "08/05/2018",
+                    TemperatureC = -13,
+                    Summary = "Freezing",
+                    TemperatureF = 9
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "09/05/2018",
+                    TemperatureC = -16,
+                    Summary = "Balmy",
+                    TemperatureF = 4
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "10/05/2018",
+                    TemperatureC = 2,
+                    Summary = "Chilly",
+                    TemperatureF = 29
+                }
+            };
+
+            public class WeatherRow
+            {
+                public string DateFormatted { get; set; }
+                public int TemperatureC { get; set; }
+                public string Summary { get; set; }
+                public int TemperatureF { get; set; }
+            }
+
+            public WeatherRow[] RowsToDisplay { get; set; }
+
+            protected override async Task OnParametersSetAsync()
+            {
+                // Simulate an async workflow.
+                await Task.Yield();
+                RowsToDisplay = _weatherData;
+            }
+
+            protected override void BuildRenderTree(RenderTreeBuilder builder)
+            {
+                base.BuildRenderTree(builder);
+                var s = 0;
+                builder.OpenElement(s++, "table");
```

Comment:
The sequence numbers here are pretty weird. They would only be like this if the developer manually unrolled the loop in their code (as in, rendered pure static content without the `foreach`).

I'd expect it to work the same if you set the sequence number to `0` everywhere and didn't increment it. Assuming that does work the same, could we express the code that way? It would be less misleading.